04-onion-routing.md

```diff
@@ -997,46 +1008,93 @@ Note that these error messages are not reliable, as they are not placed on-chain
 
 Intermediate hops store the shared secret from the forward path and reuse it to
-obfuscate any corresponding return packet during each hop.
+authenticate and obfuscate any corresponding return packet during each hop.
 In addition, each node locally stores data regarding its own sending peer in the
 route, so it knows where to return-forward any eventual return packets.
-The node generating the error message (_erring node_) builds a return packet
+
+The origin node signals to each node on the path that it supports attributable
+errors by setting the `attributable_error` tlv record in the forward packet. It
+must make sure that each node supports this format by observing the node feature
+`option_attributable_error`.
+
+## Erring node
+
+The node generating the error message builds a return packet
 consisting of the following fields:
 
 1. data:
-   * [`32*byte`:`hmac`]
    * [`u16`:`failure_len`]
    * [`failure_len*byte`:`failuremsg`]
    * [`u16`:`pad_len`]
    * [`pad_len*byte`:`pad`]
+   * [`20*5*byte`:`payloads`]
+   * [`210*4*byte`:`hmacs`]
 
-Where `hmac` is an HMAC authenticating the remainder of the packet, with a key
-generated using the above process, with key type `um`, `failuremsg` as defined
-below, and `pad` as the extra bytes used to conceal length.
 
-The erring node then generates a new key, using the key type `ammag`.
-This key is then used to generate a pseudo-random stream, which is in turn
-applied to the packet using `XOR`.
 
-The obfuscation step is repeated by every hop along the return path.
-Upon receiving a return packet, each hop generates its `ammag`, generates the
-pseudo-random byte stream, and applies the result to the return packet before
-return-forwarding it.
+Where `failuremsg` is defined below, and `pad` are the extra bytes used to
+conceal length.
 
-The origin node is able to detect that it's the intended final recipient of the
-return message, because of course, it was the originator of the corresponding
-forward packet.
-When an origin node receives an error message matching a transfer it initiated
-(i.e. it cannot return-forward the error any further) it generates the `ammag`
-and `um` keys for each hop in the route.
-It then iteratively decrypts the error message, using each hop's `ammag`
-key, and computes the HMAC, using each hop's `um` key.
-The origin node can detect the sender of the error message by matching the
-`hmac` field with the computed HMAC.
+The field `payloads` contains a per-hop payload. The erring node puts its
+payload at the start of this array and zeroes out the rest. The size of the
+field is based on the maximum supported number of hops in a route (20) and the
+payload size (5 bytes).
 
-The association between the forward and return packets is handled outside of
-this onion routing protocol, e.g. via association with an HTLC in a payment
-channel.
+The per-hop payload consists of the following fields:
+   * [`byte`:`payload_source`]
+   * [`uint32`:`hold_time_ms`]
```

Comment:
One thing I now realize is that even if there's a node on the path that messes with the data and/or hmacs, the upstream nodes will still have valid hmacs and also valid `hold_time_ms`.

Information that isn't all that useful in that case though. The sender will simply proceed with penalizing the bad node, retry, and only process timing information after "proper" failures.